src/pudl/analysis/demand_mapping.py

```diff
@@ -605,77 +605,90 @@ def allocate_and_aggregate(
-        ``respondent_id`` table, plus  a new one named ``respondent_type`` which can
-        take on the values ``balancing_authority``, ``utility``, or the special value
-        pandas.NA.
+        pandas.DataFrame: A dataframe containing 2 columns: ``eia_code`` and
+        ``respondent_type``.
 
     """
-    ba_ids = set(ba_eia861.balancing_authority_id_eia.dropna())
-    util_not_ba_ids = set(
-        utils_eia860.utility_id_eia.dropna()).difference(ba_ids)
-    new_rids = rids_ferc714.copy()
-    new_rids["respondent_type"] = pd.NA
-    new_rids.loc[new_rids.eia_code.isin(
-        ba_ids), "respondent_type"] = "balancing_authority"
-    new_rids.loc[new_rids.eia_code.isin(
-        util_not_ba_ids), "respondent_type"] = "utility"
-    ba_rids = new_rids[new_rids.respondent_type == "balancing_authority"]
-    util_rids = new_rids[new_rids.respondent_type == "utility"]
-    na_rids = new_rids[new_rids.respondent_type.isnull()]
-
-    ba_rids = (
-        ba_rids.merge(
-            ba_eia861
-            .filter(like="balancing_")
-            .drop_duplicates(subset=[
-                "balancing_authority_id_eia",
-                "balancing_authority_code_eia"]),
-            how="left", left_on="eia_code", right_on="balancing_authority_id_eia"
+    if priority == "balancing_authority":
+        primary = "balancing_authority"
+        secondary = "utility"
+    elif priority == "utility":
+        primary = "utility"
+        secondary = "balancing_authority"
+    else:
+        raise ValueError(
+            f"Invalid respondent type {priority} chosen as priority."
+            "Must be either 'utility' or 'balancing_authority'."
         )
+
+    eia_codes = pd.DataFrame(eia_codes, columns=["eia_code"]).drop_duplicates()
+    ba_ids = (
+        pd.Series(ba_ids, name="balancing_authority_id_eia")
+        .drop_duplicates()
+        .astype(pd.Int64Dtype())
     )
-    util_rids = (
-        util_rids.merge(
-            utils_eia860[["utility_id_eia", "utility_name_eia"]]
-            .drop_duplicates("utility_id_eia"),
-            how="left", left_on="eia_code", right_on="utility_id_eia"
-        )
+    util_ids = (
+        pd.Series(util_ids, name="utility_id_eia")
+        .drop_duplicates()
+        .astype(pd.Int64Dtype())
     )
-    new_rids = (
-        pd.concat([ba_rids, util_rids, na_rids])
-        .astype({
-            "respondent_type": pd.StringDtype(),
-            "balancing_authority_code_eia": pd.StringDtype(),
-            "balancing_authority_id_eia": pd.Int64Dtype(),
-            "balancing_authority_name_eia": pd.StringDtype(),
-            "utility_id_eia": pd.Int64Dtype(),
-            "utility_name_eia": pd.StringDtype(),
-        })
+
+    df = (
+        eia_codes
+        .merge(ba_ids, left_on="eia_code", right_on="balancing_authority_id_eia", how="left")
+        .merge(util_ids, left_on="eia_code", right_on="utility_id_eia", how="left")
+    )
```

Comment:
I'm not sure how this function is used in context, but the reasoning for why you do all this processing to turn lists into dfs and then merge the dfs back together seems hard to see. I assume these sets of codes were already in dfs, right? It is obviously fine to do it this way, so this is not a request to change it, but out of context it seems strange.